### PR TITLE
Add GCE Managed Instance Group target plugin

### PR DESCRIPTION
**Pull request by jsiebens**
This PR adds a Google Compute Engine MIG target plugin, for the scaling a Nomad cluster clients on GCP via manipulating [Regional Managed Instance Groups](https://cloud.google.com/compute/docs/instance-groups/regional-migs).

### CODE REVIEW

**Comment by jrasell** [comment]
Hi @jsiebens, thanks so much for this!

It will likely take us a little bit of time to test fully as we don't currently run within this provider for any of our testing. In the meantime, I hope to make a first pass review in the next day.

File: plugins/builtin/target/gce-mig/main.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/plugins"
+	"github.com/hashicorp/nomad-autoscaler/plugins/builtin/target/gce-mig/plugin"
+)
+
+func main() {
+	plugins.Serve(factory)
+}
+
+// factory returns a new instance of the AWS ASG plugin.
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/main.go:13]: copy/paste remnants?
```suggestion
// factory returns a new instance of the Google Cloud Engine MIG plugin.
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	"context"
+	"fmt"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
+	"github.com/mitchellh/go-homedir"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/option"
+	"io/ioutil"
+	"os"
+	"time"
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:13]: nit: could the imports be separated; stdlib top and other below, with the two sections separated by a newline and each section sorted alphabetically?

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+)
+
+func (t *TargetPlugin) setupGCEClients(config map[string]string) error {
+
+	credentials, ok := config[configKeyCredentials]
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:19]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	if ok {
+		contents, err := pathOrContents(credentials)
+		if err != nil {
+			return fmt.Errorf("failed to read credentials: %v", err)
+		}
+
+		t.service, err = compute.NewService(context.Background(), option.WithCredentialsJSON([]byte(contents)))
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:27]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to create Google Compute Engine client: %v", err)
+		}
+	} else {
+		service, err := compute.NewService(context.Background())
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:33]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		t.service = service
+	}
+
+	return nil
+}
+
+func (t *TargetPlugin) scaleOut(ctx context.Context, project string, zone string, mig string, num int64) error {
+
+	_, err := t.service.RegionInstanceGroupManagers.Resize(project, zone, mig, num).Context(ctx).Do()
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:47]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	// Terminate the detached instances.
+	log.Debug("deleting gce MIG instances")
+
+	request := &compute.RegionInstanceGroupManagersDeleteInstancesRequest{
+		Instances: instanceIDs,
+	}
+
+	_, err = t.service.RegionInstanceGroupManagers.DeleteInstances(project, region, mig, request).Context(ctx).Do()
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:86]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/gce.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	for _, node := range ids {
+		instanceIDs = append(instanceIDs, node.RemoteID)
+	}
+
+	// Create a logger for this action to pre-populate useful information we
+	// would like on all log lines.
+	log := t.logger.With("action", "scale_in", "instance_group", mig, "instances", ids)
+
+	// Terminate the detached instances.
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/gce.go:78]: Could this maybe be reworded? My understanding is that the instances do not need to be detached before calling terminate and that is all handled as part of the `DeleteInstances` call?

File: plugins/builtin/target/gce-mig/plugin/plugin.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	"context"
+	"fmt"
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/plugins"
+	"github.com/hashicorp/nomad-autoscaler/plugins/base"
+	"github.com/hashicorp/nomad-autoscaler/plugins/target"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/nomad"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils"
+	"google.golang.org/api/compute/v1"
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/plugin.go:13]: nit: could the imports be separated; stdlib top and other below, with the two sections separated by a newline and each section sorted alphabetically?

File: plugins/builtin/target/gce-mig/plugin/plugin_test.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
+package plugin
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/plugin_test.go:5]:
```suggestion
     "testing"
        
	"github.com/stretchr/testify/assert"
```

File: plugins/builtin/target/gce-mig/plugin/plugin.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+		err = fmt.Errorf("failed to perform scaling action: %v", err)
+	}
+	return err
+}
+
+// Status satisfies the Status function on the target.Target interface.
+func (t *TargetPlugin) Status(config map[string]string) (*sdk.TargetStatus, error) {
+
+	migRef, err := t.calculateMIG(config)
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/plugin.go:126]:
```suggestion
```

File: plugins/builtin/target/gce-mig/plugin/plugin.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+// Scale satisfies the Scale function on the target.Target interface.
+func (t *TargetPlugin) Scale(action sdk.ScalingAction, config map[string]string) error {
+
+	// GCE can't support dry-run like Nomad, so just exit.
+	if action.Count == sdk.StrategyActionMetaValueDryRunCount {
+		return nil
+	}
+
+	migRef, err := t.calculateMIG(config)
+
```
Comment (jrasell) [inline plugins/builtin/target/gce-mig/plugin/plugin.go:89]:
```suggestion
```

**Review by jrasell — CHANGES_REQUESTED** [review]
Thanks again for this @jsiebens, we are really looking forward to testing this and getting this available!

I have taken an initial pass and have made some fairly nit-picky comments/suggestions, other than those its looking good. A general comment I also noticed is that it would be nice where logging to use `GCE` over `gce` as it's an acronym.

**Review by jrasell — COMMENTED** [review]
Hi @jsiebens , we managed to get round to testing this properly in a GCP env and it worked first time without any problems, so thanks again!

One point that we noticed was that GCP has both regional and zonal managed instance groups, which use different API calls. Currently the plugin supports regional MIGs, which is likely to be the most commonly used. We would like to add support for both types, potentially making the different API calls based on whether the target includes a region or zone parameter. The team are happy to undertake this work as a follow up, but I wanted to check whether you had the desire and time to add this change?

Another, potentially trickier addition which again we can add as a follow up, would be to have some confirmation methods which ensure the change of count completes successfully. Azure has this functionality as part of the SDK, with AWS we poll the scaling status until progress reaches 100. I have not had a chance to find the GCP MIG equivalent yet though.

I'd be interested in your thoughts. Thanks again!

**Comment by jsiebens** [comment]
Hi @jrasell , thank you for the feedback, I'm glad the testing worked out well. :-)

I considered adding support for zonal MIGs, but thought it would be better to get some early feedback from the team. So here we are :-) 

I would be happy to add those changes to this PR. I'll try to find some time in the coming days and also have a closer look at the second point you mentioned.

Of course, if the team wants to have the changes sooner, don't hesitate to undertake the work.

**Review by lgfa29 — APPROVED** [review]
Thanks for the PR @jsiebens!

I tested the new zone MIGs and it all worked well 🙂  

I pushed a few small changes for some small typos and also merged `master` to fix an annoying conflict in `go.sum`.

**Pull request merged**